File: src/main/python/aoc/graph.py
#! /usr/bin/env python3

import sys
from collections import defaultdict
from collections import deque
from collections.abc import Iterator
from queue import PriorityQueue
from typing import Callable
from typing import NamedTuple
from typing import TypeVar

T = TypeVar("T")


def bfs(
    start: T,
    is_end: Callable[[T], bool],
    adjacent: Callable[[T], Iterator[T]],
) -> tuple[int, list[T]]:
    q: deque[tuple[int, T]] = deque()
    q.append((0, start))
    seen: set[T] = set()
    seen.add(start)
    parent: dict[T, T] = {}
    while not len(q) == 0:
        distance, node = q.popleft()
        if is_end(node):
            path = [node]
            curr = node
            while curr in parent:
                curr = parent[curr]
                path.append(curr)
            return distance, path
        for n in adjacent(node):
            if n in seen:
                continue
            seen.add(n)
            parent[n] = node
            q.append((distance + 1, n))
    raise RuntimeError("unsolvable")


def bfs_full(
    start: T,
    is_end: Callable[[T], bool],
    adjacent: Callable[[T], Iterator[T]],
) -> tuple[dict[T, int], dict[T, T]]:
    q: deque[tuple[int, T]] = deque()
    q.append((0, start))
    seen: set[T] = set()
    seen.add(start)
    parent: dict[T, T] = {}
    dists = defaultdict[T, int](int)
    while not len(q) == 0:
        distance, node = q.popleft()
        if is_end(node):
            dists[node] = distance
        for n in adjacent(node):
            if n in seen:
                continue
            seen.add(n)
            parent[n] = node
            q.append((distance + 1, n))
    return dists, parent


def flood_fill(
    start: T,
    adjacent: Callable[[T], Iterator[T]],
) -> set[T]:
    q: deque[T] = deque()
    q.append(start)
    seen: set[T] = set()
    seen.add(start)
    while not len(q) == 0:
        node = q.popleft()
        for n in adjacent(node):
            if n in seen:
                continue
            seen.add(n)
            q.append(n)
    return seen


class AllResults[T](NamedTuple):
    start: T
    distances: dict[T, int]
    prececessors: dict[T, list[T]]

    def get_paths(self, t: T) -> list[list[T]]:
        if t == self.start:
            return [[self.start]]
        paths = list[list[T]]()
        for predecessor in self.prececessors.get(t, list()):
            for path in self.get_paths(predecessor):
                paths.append(path + [t])
        return paths

    def get_distance(self, t: T) -> int:
        return self.distances[t]


class Dijkstra:

    @classmethod
    def dijkstra(
        cls,
        start: T,
        is_end: Callable[[T], bool],
        adjacent: Callable[[T], Iterator[T]],
        get_cost: Callable[[T, T], int],
    ) -> tuple[int, dict[T, int], list[T]]:
        q: PriorityQueue[tuple[int, T]] = PriorityQueue()
        q.put((0, start))
        best: defaultdict[T, int] = defaultdict(lambda: sys.maxsize)
        best[start] = 0
        parent: dict[T, T] = {}
        path = []
        while not q.empty():
            cost, node = q.get()
            if is_end(node):
                path = [node]
                curr = node
                while curr in parent:
                    curr = parent[curr]
                    path.append(curr)
                break
            c_total = best[node]
            for n in adjacent(node):
                new_risk = c_total + get_cost(node, n)
                if new_risk < best[n]:
                    best[n] = new_risk
                    parent[n] = node
                    q.put((new_risk, n))
        return cost, best, path

    @classmethod
    def distance(
        cls,
        start: T,
        is_end: Callable[[T], bool],
        adjacent: Callable[[T], Iterator[T]],
        get_cost: Callable[[T, T], int],
    ) -> int:
        q = PriorityQueue[tuple[int, T]]()
        q.put((0, start))
        distances = defaultdict[T, int](lambda: sys.maxsize)
        distances[start] = 0
        while not q.empty():
            distance, node = q.get()
            if is_end(node):
                return distance
            total = distances[node]
            if distance > total:
                continue
            for n in adjacent(node):
                new_distance = total + get_cost(node, n)
                if new_distance < distances[n]:
                    distances[n] = new_distance
                    q.put((new_distance, n))
        raise RuntimeError("unreachable")

    @classmethod
    def all(
        cls,
        start: T,
        is_end: Callable[[T], bool],
        adjacent: Callable[[T], Iterator[T]],
        get_cost: Callable[[T, T], int],
    ) -> AllResults[T]:
        q = PriorityQueue[tuple[int, T]]()
        q.put((0, start))
        distances = defaultdict[T, int](lambda: sys.maxsize)
        distances[start] = 0
        predecessors = dict[T, list[T]]()
        while not q.empty():
            distance, node = q.get()
            if is_end(node):
                break
            total = distances[node]
            if distance > total:
                continue
            for n in adjacent(node):
                new_distance = total + get_cost(node, n)
                dist_n = distances[n]
                if new_distance < dist_n:
                    distances[n] = new_distance
                    predecessors[n] = [node]
                    q.put((new_distance, n))
                elif new_distance == dist_n:
                    predecessors.setdefault(n, []).append(node)
        return AllResults(start, distances, predecessors)

File: src/main/rust/aoc/src/graph.rs
use std::{
    cmp::Ordering,
    collections::{BinaryHeap, HashMap, HashSet, VecDeque},
    hash::Hash,
};

#[derive(Clone, Eq, PartialEq)]
struct State<T> {
    node: T,
    distance: usize,
}

impl<T> PartialOrd for State<T>
where
    T: Eq + Copy + Hash,
{
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl<T> Ord for State<T>
where
    T: Eq + Copy + Hash,
{
    fn cmp(&self, other: &Self) -> Ordering {
        other.distance.cmp(&self.distance)
    }
}

pub struct BFS<T>(T);

impl<T> BFS<T>
where
    T: Eq + Copy + Hash,
{
    pub fn execute(
        start: T,
        is_end: impl Fn(T) -> bool,
        adjacent: impl Fn(T) -> Vec<T>,
    ) -> usize {
        let mut q: VecDeque<State<T>> = VecDeque::new();
        q.push_back(State {
            node: start,
            distance: 0,
        });
        let mut seen: HashSet<T> = HashSet::new();
        seen.insert(start);
        while let Some(state) = q.pop_front() {
            if is_end(state.node) {
                return state.distance;
            }
            adjacent(state.node).iter().for_each(|n| {
                if !seen.contains(n) {
                    seen.insert(*n);
                    q.push_back(State {
                        node: *n,
                        distance: state.distance + 1,
                    });
                }
            });
        }
        unreachable!();
    }

    pub fn flood_fill(start: T, adjacent: impl Fn(T) -> Vec<T>) -> HashSet<T> {
        let mut q: VecDeque<T> = VecDeque::new();
        q.push_back(start);
        let mut seen: HashSet<T> = HashSet::new();
        seen.insert(start);
        while let Some(node) = q.pop_front() {
            adjacent(node).iter().for_each(|n| {
                if !seen.contains(n) {
                    seen.insert(*n);
                    q.push_back(*n);
                }
            });
        }
        seen
    }

    pub fn flood_fill_mut(
        start: T,
        mut adjacent: impl FnMut(T) -> Vec<T>,
    ) -> HashSet<T> {
        let mut q: VecDeque<T> = VecDeque::new();
        q.push_back(start);
        let mut seen: HashSet<T> = HashSet::new();
        seen.insert(start);
        while let Some(node) = q.pop_front() {
            adjacent(node).iter().for_each(|n| {
                if !seen.contains(n) {
                    seen.insert(*n);
                    q.push_back(*n);
                }
            });
        }
        seen
    }
}

pub struct AStar<T>(T);

pub struct Result<T> {
    start: T,
    distances: HashMap<T, usize>,
    paths: HashMap<T, T>,
}

pub struct AllResults<T> {
    start: T,
    predecessors: HashMap<T, Vec<T>>,
}

impl<T> Result<T>
where
    T: Eq + Copy + Hash,
{
    pub fn get_distance(&self, t: T) -> Option<usize> {
        self.distances.get(&t).cloned()
    }

    pub fn get_distances(&self) -> &HashMap<T, usize> {
        &self.distances
    }

    pub fn get_path(&self, t: T) -> Option<Vec<T>> {
        match self.get_distance(t) {
            None => None,
            Some(_) => {
                let mut path = vec![];
                let mut parent = Some(t);
                let source = self.start;
                if t != self.start {
                    while parent.is_some() && parent.unwrap() != self.start {
                        let p = parent.unwrap();
                        path.insert(0, p);
                        parent = self.paths.get(&p).copied()
                    }
                    path.insert(0, source);
                } else {
                    path.push(source);
                }
                Some(path)
            }
        }
    }
}

impl<T> AllResults<T>
where
    T: Eq + Copy + Hash,
{
    pub fn get_paths(&self, t: T) -> Vec<Vec<T>> {
        if t == self.start {
            return vec![vec![self.start]];
        }
        let mut paths = vec![];
        for predecessor in self.predecessors.get(&t).unwrap_or(&Vec::new()) {
            for mut path in self.get_paths(*predecessor) {
                path.push(t);
                paths.push(path);
            }
        }
        paths
    }
}

impl<T> AStar<T>
where
    T: Eq + Copy + Hash,
{
    pub fn execute(
        start: T,
        is_end: impl Fn(T) -> bool,
        adjacent: impl Fn(T) -> Vec<T>,
        cost: impl Fn(T) -> usize,
    ) -> Result<T> {
        let mut q: BinaryHeap<State<T>> = BinaryHeap::new();
        q.push(State {
            node: start,
            distance: 0,
        });
        let mut distances: HashMap<T, usize> = HashMap::new();
        distances.insert(start, 0);
        let mut paths: HashMap<T, T> = HashMap::new();
        while let Some(state) = q.pop() {
            if is_end(state.node) {
                break;
            }
            let total = *distances.get(&state.node).unwrap_or(&usize::MAX);
            if state.distance > total {
                continue;
            }
            adjacent(state.node).iter().for_each(|n| {
                let risk = total + cost(*n);
                if risk < *distances.get(n).unwrap_or(&usize::MAX) {
                    distances.insert(*n, risk);
                    paths.insert(*n, state.node);
                    q.push(State {
                        node: *n,
                        distance: risk,
                    });
                }
            });
        }
        Result {
            start,
            distances,
            paths,
        }
    }

    pub fn all(
        start: T,
        is_end: impl Fn(T) -> bool,
        adjacent: impl Fn(T) -> Vec<T>,
        cost: impl Fn(T, T) -> usize,
    ) -> AllResults<T> {
        let mut q: BinaryHeap<State<T>> = BinaryHeap::new();
        q.push(State {
            node: start,
            distance: 0,
        });
        let mut distances: HashMap<T, usize> = HashMap::new();
        distances.insert(start, 0);
        let mut predecessors: HashMap<T, Vec<T>> = HashMap::new();
        while let Some(state) = q.pop() {
            if is_end(state.node) {
                break;
            }
            let total = *distances.get(&state.node).unwrap_or(&usize::MAX);
            if state.distance > total {
                continue;
            }
            adjacent(state.node).iter().for_each(|n| {
                let risk = total + cost(state.node, *n);
                let dist_n = *distances.get(n).unwrap_or(&usize::MAX);
                match risk.cmp(&dist_n) {
                    Ordering::Less => {
                        distances.insert(*n, risk);
                        predecessors.entry(*n).insert_entry(vec![state.node]);
                        q.push(State {
                            node: *n,
                            distance: risk,
                        });
                    }
                    Ordering::Equal => {
                        predecessors
                            .entry(*n)
                            .and_modify(|e| e.push(state.node))
                            .or_insert(vec![state.node]);
                    }
                    _ => (),
                }
            });
        }
        AllResults {
            start,
            predecessors,
        }
    }

    pub fn distance(
        start: T,
        is_end: impl Fn(T) -> bool,
        adjacent: impl Fn(T) -> Vec<T>,
        cost: impl Fn(T, T) -> usize,
    ) -> usize {
        let mut q: BinaryHeap<State<T>> = BinaryHeap::new();
        q.push(State {
            node: start,
            distance: 0,
        });
        let mut distances: HashMap<T, usize> = HashMap::new();
        distances.insert(start, 0);
        while let Some(state) = q.pop() {
            if is_end(state.node) {
                return state.distance;
            }
            let total = *distances.get(&state.node).unwrap_or(&usize::MAX);
            if state.distance > total {
                continue;
            }
            adjacent(state.node).iter().for_each(|n| {
                let risk = total + cost(state.node, *n);
                if risk < *distances.get(n).unwrap_or(&usize::MAX) {
                    distances.insert(*n, risk);
                    q.push(State {
                        node: *n,
                        distance: risk,
                    });
                }
            });
        }
        panic!("unsolvable");
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::grid::{Cell, CharGrid, Grid};

    fn bfs_execute(start: Cell, end: Cell) -> usize {
        #[rustfmt::skip]
        let v = &vec![
            ".###",
            "..##",
            "#..#",
            "##..",
            "###."];
        let grid = CharGrid::from(&v);
        let adjacent = |cell| {
            grid.capital_neighbours(&cell)
                .iter()
                .filter(|n| grid.get(&n) != '#')
                .cloned()
                .collect()
        };
        BFS::execute(start, |cell| cell == end, adjacent)
    }

    #[test]
    pub fn bfs_execute_ok() {
        assert_eq!(bfs_execute(Cell::at(0, 0), Cell::at(4, 3)), 7);
        assert_eq!(bfs_execute(Cell::at(2, 2), Cell::at(0, 0)), 4);
    }

    #[test]
    #[should_panic]
    pub fn bfs_unreachable() {
        bfs_execute(Cell::at(0, 0), Cell::at(4, 0));
    }

    #[test]
    pub fn bfs_flood_fill() {
        #[rustfmt::skip]
        let v = &vec![
            ".###",
            "..##",
            "#..#",
            "##..",
            "###."];
        let grid = CharGrid::from(&v);
        let adjacent = |cell| {
            grid.capital_neighbours(&cell)
                .iter()
                .filter(|n| grid.get(&n) != '#')
                .cloned()
                .collect()
        };
        assert_eq!(
            BFS::flood_fill(Cell::at(2, 2), adjacent),
            HashSet::from([
                Cell::at(0, 0),
                Cell::at(1, 0),
                Cell::at(1, 1),
                Cell::at(2, 1),
                Cell::at(2, 2),
                Cell::at(3, 2),
                Cell::at(3, 3),
                Cell::at(4, 3),
            ])
        );
    }

    #[test]
    pub fn astar_execute_ok() {
        #[rustfmt::skip]
        let v = &vec![
            ".###",
            "..##",
            "#..#",
            "##..",
            "###."];
        let grid = CharGrid::from(&v);
        let adjacent = |cell| {
            grid.capital_neighbours(&cell)
                .iter()
                .filter(|n| grid.get(&n) != '#')
                .cloned()
                .collect()
        };
        let result = AStar::execute(
            Cell::at(0, 0),
            |cell| cell == Cell::at(4, 3),
            adjacent,
            |_| 1,
        );
        assert_eq!(result.get_distance(Cell::at(2, 2)).unwrap(), 4);
        assert_eq!(result.get_distance(Cell::at(4, 3)).unwrap(), 7);
        assert_eq!(
            result.get_path(Cell::at(2, 2)).unwrap(),
            vec![
                Cell::at(0, 0),
                Cell::at(1, 0),
                Cell::at(1, 1),
                Cell::at(2, 1),
                Cell::at(2, 2)
            ]
        );
        assert!(result.get_distance(Cell::at(4, 0)).is_none());
        assert!(result.get_path(Cell::at(4, 0)).is_none());
    }
}
